### ISSUE TITLE
ccache -s cache size doesn't correspond with folder size

### ISSUE THREAD

**Issue by sp-francisco-sanchez** [ccache/ccache#124]
Hi!

I've been experimenting with ccache in a network share using 4 build machines and I've observed two main things:
- The cache size that is shown when executing the stats command doesn't correspond with the real size of the ccache dir.
- The stats change with other values without any explanation since we are never executing a zero stats/clear command.

The cache size from the stats command was showing around [1.5/50.0] GB while the real weight of the ccache dir was 265GB!!!
After executing a [ccache -c] command everything returned to normality but... during the process the cache size shown during the [ccache -s] command was increased to [41.5/50.0] GB from the initial [1.5/50.0] GB and the real weight of the ccache dir was reduced to 41.5 GB that is what is being shown in the [stats -s] command...

I have used the same .conf file across the machines, proper set the umask parameter, the network share and I've run the chmod command as shown in the documentation...

What can be causing this behavior?

**Comment by jrosdahl** [comment]
Sounds like the locking to protect the "stats files" (which keep track of cache size and other statistics like hit/miss rate) doesn't work in your setup.

What do you mean with "network share"? Which network protocol (file system) is used? Are you on Windows or some Unix system?

**Comment by sp-francisco-sanchez** [comment]
Hi!

Thank you very much for your prompt response. That's what I thought too but I think I have everything correctly setup.
It is a NFS filesystem mounted in a Mac OS El Capitan machine.

The /etc/exports entry for the shared folder is: (the ${my_user} is the username and not a variable and it exists in all the machines)

/Users/${my_user}/ccache_share -mapall=${my_user} ${machine_list}

I hope it can help...

**Comment by jrosdahl** [comment]
I just whipped together a quick and crude program for stress testing ccache's lockfile code:

http://joel.rosdahl.net/lockfile-stress-test.c

Start by trying it on only one of your hosts:

1. Compile it, e.g. like this: `gcc lockfile-stress-test.c -o lockfile-stress-test`
2. `echo 0 >x` (where x is a path to a file on the same file system as the ccache directory)
3. `./lockfile-stress-test 4 1000 x`

The first argument is the number of parallel jobs. The second argument is the number of times to increase a counter. The third argument is the file storing the counter. You can try tweaking the numbers if it takes too long time to run on your setup, but the number of parallel jobs should of course be more than 1.

The output looks like this for me on a local file system:

```
lockfile_acquire: could not acquire x.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 32000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 32000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 64000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 64000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 32000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 128000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire x.lock; sleeping 32000 us before retrying
The counter should now have been increased by 4000
Actual counter before execution: 0
Actual counter after execution: 4000
```

If the counter increase looks OK for your NFS mount when running on only one host, try starting the program in parallel on different hosts to see if they can update the counter file without overwriting each other.

**Comment by sp-francisco-sanchez** [comment]
Hi!

Thank you very much for all the information and implication.
I paste here the output. I setup a new NFS in the /etc/exports file with the same configuration as the one shown above.
I mount also the newly created NFS share in the same machine.

I use a SSD.

EXECUTION IN LOCAL DIRECTORY
If I execute the tool you supplied locally in the directory I use for the network share I get no problems when I run it with 4 parallel jobs. All the acquires are fine.
If I increase the parallel jobs to 8, then I get sometimes assertions and I don't get the file increased in 1000 units.

EXECUTION IN THE MOUNT POINT
If I execute the tool you supplied using the file in the mount point (but from the same machine the directory is exported from) even with only 2 threads I get always some asserts:
`lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 32000 us before retrying
Failed to acquire lock ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_job, file $HOME/ccache_test/lockfile-stress-test.c, line 151.
The counter should now have been increased by 2000
Actual counter before execution: 0
Actual counter after execution: 1000`

EXECUTION IN PARALLEL FROM 2 MACHINES:
I always get several assertions in both sides:

MACHINE A => `lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
Failed to acquire lock ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_job, file $HOME/ccache_test/lockfile-stress-test.c, line 151.
Failed to acquire lock ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock
Failed to acquire lock ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_joAssertion failed: (acquired), fub, file $HOME/ccachnction start_increase_counter_joe_test/lockfile-stress-test.c, lb, file $HOME/ccachine 151.
e_test/lockfile-stress-test.c, line 151.
Assertion failed: (ret == 0), function write_counter, file $HOME/ccache_test/lockfile-stress-test.c, line 137.
The counter should now have been increased by 4000
Actual counter before execution: 153
Actual counter after execution: 160`

MACHINE B => `lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
Failed to acquire lock ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_job, file $HOME/ccache_test/lockfile-stress-test.c, line 151.
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 8000 us before retrying
Assertion failed: (ret == 0), function write_counter, file $HOME/ccache_test/lockfile-stress-test.c, line 137.
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
Failed to acquire lock ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_job, file $HOME/ccache_test/lockfile-stress-test.c, line 151.
The counter should now have been increased by 4000
Actual counter before execution: 0
Actual counter after execution: 1009`


If I start both programs at almost the same time (after having done this several times) I got once this output in both machines in a never-ending loop. It never finished having to cancel the execution in both machines. It didn't matter I cancelled the program in one of the machines, the other showed exactly the same behavior when I left it running alone afterwards:
`
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 4096000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 4096000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 8192000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 8192000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 8192000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 8192000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 16384000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 16384000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 16384000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 16384000 us before retrying
lockfile_acquire: breaking /Users/buildmaster/fury_test_nfs/test_lock.txt.lock
lockfile_acquire: breaking /Users/buildmaster/fury_test_nfs/test_lock.txt.lock
lockfile_acquire: breaking /Users/buildmaster/fury_test_nfs/test_lock.txt.lock
lockfile_acquire: breaking /Users/buildmaster/fury_test_nfs/test_lock.txt.lock
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 4000 us before retrying
`

Any clue what can be causing this behavior?

Thanks in advance!

**Comment by jrosdahl** [comment]
My gut feeling is bad NFS configuration or bad NFS implementation, but it's hard to say exactly what. And I'm by no means an NFS expert.

I have now added some more printouts to the program (at the previously mentioned URL), so you could try to recompile and rerun it to hopefully get more info when the lock can't be acquired.

**Comment by afbjorklund** [comment]
We had (and have) similar problems with NFS. There's also the problem where ccache doesn't lock the cache during the clean operation, so if anything is added during it is not added to the statistics. The workaround was running "ccache -s" from a cron job, and disabling clean from individual clients. This also helped with the ddos attack that happens when everyone starts cleaning the cache simultaneously...

**Comment by afbjorklund** [comment]
@jrosdahl : here was one experiment, that tried to "prime" the stats cache and then modify the statistics incrementally instead of resetting them afterwards: 403cb007c751ed59987148674716f3061c5f1079
But in the end we had to give up on NFS as primary cache anyway (for performance reasons) and I'm not sure that using NFS as secondary cache is a better alternative either - compared to e.g. memcached ?

**Comment by sp-francisco-sanchez** [comment]
Hi! Thank you very much for your interest.

After testing it in two machines at the same time I get almost the same results. Iterating over and over finally I got something different. See the output:

`lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: readlink failed after symlink succeeded: Stale NFS file handle
Failed to acquire lock /Users/buildmaster/fury_test_nfs/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_job, file /Users/buildmaster/ccache_test/lockfile-stress-test.c, line 154.
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 32000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 64000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 64000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 128000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 128000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 256000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 256000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 512000 us before retrying
Assertion failed: (ret == 0), function write_counter, file /Users/buildmaster/ccache_test/lockfile-stress-test.c, line 140.
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 1024000 us before retrying
lockfile_acquire: could not acquire /Users/buildmaster/fury_test_nfs/test_lock.txt.lock; sleeping 2048000 us before retrying
lockfile_acquire: readlink failed after symlink succeeded: Stale NFS file handle
Failed to acquire lock /Users/buildmaster/fury_test_nfs/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_job, file /Users/buildmaster/ccache_test/lockfile-stress-test.c, line 154.
Assertion failed: (f), function read_counter, file /Users/buildmaster/ccache_test/lockfile-stress-test.c, line 122.
Abort trap: 6`

In the documentation appears that should be possible to use it in a shared folder but I see that there are some problems with that. Have you used ccache in a network share successfully?

Additionally I have set-up the temporary directory locally in each of the machines so it does not use the CCACHE_DIR but I still get the same results.

**Comment by sp-francisco-sanchez** [comment]
Hi! Thank you very much for your interest.

After testing it in two machines at the same time I get almost the same results. Iterating over and over finally I got something different. See the output:

lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 8000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 16000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 32000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 32000 us before retrying
Assertion failed: (ret == 0), function write_counter, file $HOME/ccache_test/lockfile-stress-test.c, line 140.
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 64000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 64000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 128000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 8000 us before retrying
lockfile_acquire: readlink failed after symlink succeeded: Stale NFS file handle
Failed to acquire lock ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock
Assertion failed: (acquired), function start_increase_counter_job, file $HOME/ccache_test/lockfile-stress-test.c, line 154.
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 256000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 1000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 2000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 4000 us before retrying
lockfile_acquire: could not acquire ${MY_SHARED_NFS_FOLDER}/test_lock.txt.lock; sleeping 8000 us before retrying
Assertion failed: (ret == 0), function write_counter, file $HOME/ccache_test/lockfile-stress-test.c, line 140.
The counter should now have been increased by 4000
Actual counter before execution: 0
Actual counter after execution: 1020

In the documentation appears that should be possible to use it in a shared folder but I see that there are some problems with that. Have you used ccache in a network share successfully?

Additionally I have set-up the temporary directory locally in each of the machines so it does not use the CCACHE_DIR but I still get the same results.

**Comment by jrosdahl** [comment]
> lockfile_acquire: readlink failed after symlink succeeded: Stale NFS file handle

Looks like a typical NFS configuration or implementation problem.

> In the documentation appears that should be possible to use it in a shared folder but I see that there are some problems with that.

Yes. Note though that the documentation says "ccache hasn't been tested very thoroughly on NFS".

> Have you used ccache in a network share successfully?

I have never used it myself, only applied patches from people that have tried using NFS. I personally quite sceptical against NFS since there often are problems.

**Comment by jrosdahl** [comment]
> There's also the problem where ccache doesn't lock the cache during the clean operation, so if anything is added during it is not added to the statistics. The workaround was running "ccache -s" from a cron job, and disabling clean from individual clients. This also helped with the ddos attack that happens when everyone starts cleaning the cache simultaneously...

Yeah, that's of course a problem even on a single machine with multiple jobs. It would be nice to have something about the cron job workaround in the documentation.

**Comment by blong3748** [comment]
We use ccache in a nfs environment with hundreds of processors beating on it and it works quite well. It is essential to us.

Re lock times, on a recent build I instrumented to parse times out of ccache log files I  saw
Num locks 1438
  min     2.00 ms
  max   410.91 ms
  avg     5.71 ms
  num over 100ms 4

Though we do get stats file lock timeouts

If the stats files were spread across 256 (if cache depth is >=2) then congestion would reduce by 16x over always storing them one level down.

The time to generate the stats report for 256 vs 16 would not be burdensome since one does not ask for stats reports too often anyway

**Comment by jrosdahl** [comment]
@sp-francisco-sanchez: Can we close this issue or do you want more feedback? If so, what?

**Comment by jrosdahl** [comment]
> If the stats files were spread across 256 (if cache depth is >=2) then congestion would reduce by 16x over always storing them one level down.

I wrote #168 to follow up on this idea.
